src/server-glue/webresource-handlers/S3Handler.ts

```diff
@@ -0,0 +1,105 @@
+import { optionalVar, requiredVar } from '@balena/env-parsing';
+import {
+	IncomingFile,
+	UploadResponse,
+	WebResourceHandler,
+} from '../webresource-handler';
+import {
+	S3Client,
+	S3ClientConfig,
+	DeleteObjectCommand,
+} from '@aws-sdk/client-s3';
+import { Upload } from '@aws-sdk/lib-storage';
+import { v4 as uuidv4 } from 'uuid';
+
+export interface S3HandlerConfig {
+	bucket?: string;
+	region?: string;
+	accessKeyId?: string;
+	secretAccessKey?: string;
+	endpoint?: string;
+}
+
+export class S3Handler implements WebResourceHandler {
+	private readonly bucket: string;
+	private readonly region: string;
+	private readonly accessKeyId: string;
+	private readonly secretAccessKey: string;
+	private readonly endpoint: string;
+	private client: S3Client;
+
+	constructor(config?: S3HandlerConfig) {
+		this.bucket =
+			config?.bucket ??
+			optionalVar('S3_STORAGE_ADAPTER_BUCKET', 'balena-pine-web-resources');
+		this.region =
+			config?.region ?? optionalVar('S3_STORAGE_ADAPTER_REGION', 'us-east-1');
```

Comment:
Both also don't have default vars as neither the bucket nor the region can be defaulted.
If they are not specified in the config nor in the env the startup should fail